### DISCUSSION TITLE
AUTOLOAD issue for non-existing key

### DISCUSSION THREAD

**Discussion by oodler577** [haukex/Util-H2O#22] [Q&A] [answered]
Hi @haukex - I am trying to add something in `Util::H2O::More` that will allow me to call a potentially non-existing key, and rather than get an error because it does not exist, simply have it return `undef`. My attempt using purely `Util::H2O` has me a little baffled and I wanted to see if it was a bug or intentional - if the latter, how should I do what I want?

```
#!/usr/bin/env perl

use v5.12;
use strict;
use warnings;

use Util::H2O qw/h2o/;

my $hash = h2o -nolock, {
  thirteen => 13,
  AUTOLOAD => sub {
    my $self = shift;
    our $AUTOLOAD;
    ( my $key = $AUTOLOAD ) =~ s/.*:://;
    $self->{$key} = shift if @_;
    return (exists $self->{$key}) ? $self->{$key} : undef;
  },
};

say $hash->twelve; # unexpectedly prints a CODE ref address

$hash->twelve(12); # works as expected

say $hash->twelve; # works as expected
```

Results in,

```
CODE(0x56209e619a28)
12
```

Also, it seems `-nolock` doesn't need to be there (or doesn't matter) - I am using it based on your example in the POD. Any ideas how to get just the `undef` instead of the CODE ref address?

**Comment by haukex** [comment, marked as answer]
Hi,

The key piece from the [documentation](https://metacpan.org/pod/Util::H2O#AUTOLOAD) is this:

> Without the `-meth` option, you will get a "catch-all" accessor to which all method calls to unknown method names will go, and with `-meth` enabled (which is implied by `-classify`), you can install your own custom AUTOLOAD handler by passing a coderef as the value for this key

Since in your example, you don't have `-meth` turned on, `$hash->anything_except_thirteen` simply acts like a catch-all storage location - you just happened to have stored a code ref in it initially. I assume you'll get what you want when you turn `-meth` on.

Hope this helps!

**Comment by oodler577** [comment]
@haukex - thank you that seems to have done it. This is doing what I want,

```
#!/usr/bin/env perl

use v5.12;
use strict;
use warnings;

use Util::H2O qw/h2o/;

my $hash = h2o -nolock, -meth, {
  thirteen => 13,
  AUTOLOAD => sub {
    my $self = shift;
    our $AUTOLOAD;
    ( my $key = $AUTOLOAD ) =~ s/.*:://;
    $self->{$key} = shift if @_;
    return $self->{$key};
  },
};

if (not $hash->twelve) {
  say "'twelve' is not set yet ... setting";
  $hash->twelve(12); # works as expected
}

say $hash->twelve; # works as expected
```

Output:

```
'twelve' is not set yet ... setting
12
```

I will close the issue, ty!

**Comment by oodler577** [comment]
@haukex I went back and forth on making this default behavior or not, but ended up making it so that `d2o` has a flag that must enable this new capability - `-autoundef` - https://metacpan.org/release/OODLER/Util-H2O-More-0.3.5/view/lib/Util/H2O/More.pm#-autoundef

I ended up borrowing you technique for parsing the flags, though I am just looking for 1. If I add more, I'll have to use your `while`.

Thanks again for the pointer in the right direction.
